### ISSUE TITLE
typing.no_type_check is not understood

### ISSUE THREAD

**Issue by earwig** [microsoft/pyright#1448]
**Describe the bug**
[@typing.no_type_check](https://docs.python.org/3/library/typing.html#typing.no_type_check) is used to decorate functions that use annotations for other purposes than type hints. I have a code base that uses special annotations for certain functions (part of a DSL) and normal type annotations for other functions.

**To reproduce**
```python
import typing

def foo(a: int) -> int:
    return "hello"

@typing.no_type_check
def bar(a: int) -> int:
    return "hello"
```

mypy handles this correctly, only complaining about `foo`:

```
example.py:4: error: Incompatible return value type (got "str", expected "int")
        return "hello"
               ^
Found 1 error in 1 file (checked 1 source file)
```
pyright additionally complains about the decorator itself and does not ignore the type annotations in `bar`:

```
  4:12 - error: Expression of type "Literal['hello']" cannot be assigned to return type "int"
    "Literal['hello']" is incompatible with "int" (reportGeneralTypeIssues)
  6:2 - error: Expected no arguments to "no_type_check" constructor (reportGeneralTypeIssues)
  8:12 - error: Expression of type "Literal['hello']" cannot be assigned to return type "int"
    "Literal['hello']" is incompatible with "int" (re
```

**Expected behavior**
`@typing.no_type_check` should cause type annotations for the decorated function to be ignored.

**VS Code extension or command-line**
Example above generated with CLI pyright 1.1.108, mypy 0.782, and Python 3.7, though I primarily use pyright through vim.

**Additional context**
You will notice that typeshed's typing.pyi defines `no_type_check = object()`, which might at first blush make this look like a bug in typeshed, but that is just typeshed's convention for marking special names that need explicit support in the typechecker. See https://github.com/python/mypy/pull/645.

**Comment by erictraut** [comment]
I intentionally don't support `no_type_check`. There are many ways to address type errors, and all of them are better than disabling type checking for entire functions. 

Can you explain why you're using `no_type_check`? From the sample above, it looks like you're implying that your code generator is emitting code that is incorrect from a type perspective, so you want to suppress type checking for that incorrect code. Why not just fix the code instead?

**Comment by earwig** [comment]
`no_type_check` is intended for functions where the annotations are not actual type hints but something else entirely and should not be checked by a type checker. It's different from cases where the types are wrong or underspecified.

I tried to explain this in my comment: "I have a code base that uses special annotations for certain functions (part of a DSL) and normal type annotations for other functions." I admit that's not a great explanation. I made up a simple example to show the error by itself, not what actual code using this decorator should look like.

Here is a hopefully clearer use case: a decorator to define an API that validates its arguments using [JSON Schema](https://json-schema.org/).

```python
def api(func):
    @functools.wraps(func)
    def inner(*args, **kwargs):
        ... # Validate args/kwargs against func.__annotations__
        ret = func(*args, **kwargs)
        ... # Validate ret against func.__annotations__['return']
        return ret
    return inner

@api
@typing.no_type_check
def foo(arg: {"description": "The foo argument; must be alphanumeric",
              "type": "string",
              "pattern": "^[a-zA-Z]*$"}
        ) -> {"type": "string", "maxLength": 8}:
    return arg[:8]
```

To be clear, this is part of a codebase that uses normal type hints for most functions. I want the benefit of mypy/pyright for everything _except_ my special functions.

IMO this use case should be supported by any type checker for general Python code. [PEP 3107](https://www.python.org/dev/peps/pep-3107/): "By itself, Python does not attach any particular meaning or significance to annotations. Left to its own, Python simply makes these expressions available as described in Accessing Function Annotations below. The only way that annotations take on meaning is when they are interpreted by third-party libraries. These annotation consumers can do anything they want with a function's annotations."

There are some other ways to address this problem, like wrapping everything with [typing.Annotated](https://docs.python.org/3/library/typing.html#typing.Annotated) (which is not available in my version of Python). If we are not going for feature parity with mypy, at the very least I suggest a clear error be emitted here.

**Comment by erictraut** [comment]
I don't think this is a compelling enough argument to add support for `no_type_check`.

`Annotated` was designed for this purpose. Please use that instead. You can enclose your annotations in quotes or use `from __future__ import annotations` or use `typing_extensions.Annotated` to work around the fact that your current version of Python doesn't support `typing.Annotated`.

**Comment by erictraut** [comment]
Another solution is to generate type stubs (".pyi" files) alongside these generated ".py" files and provide proper type annotations in the stubs.

If those suggestions don't work for you, mypy is still an option for your source base.

**Comment by earwig** [comment]
That will work. Thanks for your time.

**Comment by SandyChapman** [comment]
@erictraut I understand this was closed as designed, but this behaviour does run counter to [PEP 484](https://www.python.org/dev/peps/pep-0484/#compatibility-with-other-uses-of-function-annotations). This would specifically be useful for code prior to Python 3.9 as `Annotated` is not in the `typing` module [PEP 593](https://www.python.org/dev/peps/pep-0593/) unless you import it from the `typing_extensions` package.

> To mark portions of the program that should not be covered by type hinting, you can use one or more of the following:
> - a `# type: ignore` comment;
> - a `@no_type_check` decorator on a class or function;
> - a custom class or function decorator marked with `@no_type_check_decorator`.

**Comment by JelleZijlstra** [comment]
Why not use `from typing_extensions import Annotated`?

**Comment by SandyChapman** [comment]
Because I may not want to use `Annotated` and rely on the more simple decorator. The point I'm trying to raise is that the behaviour currently runs counter to PEP484 which Pyright explicitly calls out it supports on the main README. This is inconsistent when it appears the goal is to have Pyright be a standards-compliant typechecker.

**Comment by amacfie** [comment]
> The Python type system has evolved significantly since PEP 484 was written. `Annotated` was added to support this use case.

[The docs](https://docs.python.org/3/library/typing.html#typing.no_type_check) don't mention that `no_type_check` is deprecated. As long as it's part of Python I see value in supporting it for those who want to use it.

**Comment by erictraut** [comment]
Support for `@no_type_check` is included in pyright 1.1.347, which I just published. It will also be included a future release of pylance.
